rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileDownloader.java

```diff
@@ -94,13 +104,32 @@ private void initializeMetrics()
       log.debug("Processing Request for File : " + path.toString() + " LocalFile : " + localPath);
       ByteBuffer directWriteBuffer = bufferPool.getBuffer(diskReadBufferSize);
 
+      // Setup file metadata if not there
+      try {
+      bookKeeper.getCacheStatus(new CacheStatusRequest(
+              context.getRemoteFilePath(),
+              context.getFileSize(),
+              context.getLastModifiedTime(),
+              0L,
+              0L));
+      }
+      catch (Exception e) {
+        log.warn("Error communicating with bookKeeper", e);
+        // Exception is not expected as RemoteFetchProcessor ensures to not start processing until BookKeeper has initialized
+        // recover from this, requeue the requests for this file and continue with next file
+        remoteFetchProcessor.addToProcessQueueSafe(context.getRemoteFilePath(), context.getRanges().asRanges(), context.getFileSize(), context.getLastModifiedTime());
+        continue;
+      }
+
       FileDownloadRequestChain requestChain = new FileDownloadRequestChain(bookKeeper, fs, localPath,
           directWriteBuffer, conf, context.getRemoteFilePath(), context.getFileSize(),
           context.getLastModifiedTime());
 
       for (Range<Long> range : context.getRanges().asRanges()) {
-        ReadRequest request = new ReadRequest(range.lowerEndpoint(), range.upperEndpoint(),
-            range.lowerEndpoint(), range.upperEndpoint(), null, 0, context.getFileSize());
+        // align range to block boundary
+        long start = (range.lowerEndpoint() / blockSize) * blockSize;
+        long end = Math.min((((range.upperEndpoint() - 1) / blockSize) + 1) * blockSize, context.getFileSize());
```

Comment:
will `ceil` be more accurate?

Reply:
this is simplified in later commits but not sure how ceil would help here.

Reply:
oh, I guess I misunderstood. `end` (or `toEndBlock` with all changes) gives the first block which doesn't contain the end position right - since it used as an open end to the range and not close. is that right?

Reply:
yes, name `toEndBlock` is also not be best one. I will see if I can do better to avoid this confusion

Reply:
Added comments to the utility methods

Reply:
ok, it can be taken up later too